### DISCUSSION TITLE
Clear Formatting

### DISCUSSION THREAD

**Discussion by upperhouse** [awcodes/filament-tiptap-editor#382] [Ideas]
Hi there.
Do you have any plans of rolling in the ability to remove inline formatting within the editor. Would be great for clients who insist on pasting from Word. It looks like maybe this exists? [https://tiptap.dev/docs/editor/api/commands/clear-content](https://tiptap.dev/docs/editor/api/commands/clear-content)

**Comment by awcodes** [comment]
Clear content resets the field to an empty state. It doesn't clear formatting. And there is an erase button on the right that already calls that command.

**Comment by rudiboot** [comment]
Hi, to clarify from upperhouse question, is there a way of removing the inline style when pasting text (ie maybe just keep < p > or < h1 > without any inline style css) either when it's being pasted or after using a menu button?

**Reply by awcodes** [reply]
To my knowledge there is not. If ProseMirror or Tiptap allow attributes on the element then they will be pasted in. Any non supported attributes will be stripped out.

**Reply by rudiboot** [reply]
Ok fair enough.
Thank you for your prompt response and all the great work to make this and curator work a treat with filament

**Comment by 000yuying** [comment]
Here's the translation:

We've done something similar here by extending the document and adding buttons for reference.

The main action is through blade.php, with the action editor().chain().focus().clearNodes().unsetAllMarks().run().

clear.blade.php:
<x-filament-tiptap-editor::button
    label="Clear Formatting"
    action="editor().chain().focus().clearNodes().unsetAllMarks().run()"
>
    <svg xmlns="http://www.w3.org/2000/svg"
         width="20" height="20" viewBox="0 0 24 24">
        <path fill="currentColor" d="M12.651 14.065L11.605 20H9.574l1.35-7.661-7.41-7.41L4.93 3.515 20.485 19.07l-1.414 1.414-6.42-6.42zm-.878-6.535l.27-1.53h-1.8l-2-2H20v2h-5.927L13.5 9.257 11.773 7.53z"/>
    </svg>
    <span class="sr-only">Clear Formatting</span>
</x-filament-tiptap-editor::button>


extensions.js:
import { clearMarks } from "./clear.js";
window.TiptapEditorExtensions = {
    clearMarks: {
        clearMarks
    },
};


clear.js:
import { setTextAlign } from './text_align';
import { setTextLineHeight } from './line_height';
import CustomStrike from "./paste.js";

const FormatMarks = {
    bold: 'bold',
    italic: 'italic',
    underline: 'underline',
    strike: 'strike',
    link: 'link',
    textColor: 'text_color',
    textHighlight: 'text_highlight',
};

const FORMAT_MARK_NAMES = [
    FormatMarks.bold,
    FormatMarks.italic,
    FormatMarks.underline,
    FormatMarks.strike,
    FormatMarks.link,
    FormatMarks.textColor,
    FormatMarks.textHighlight,
];

function clearMarks(tr, schema) {
    const { doc, selection } = tr;
    if (!selection || !doc) return tr;

    const { from, to, empty } = selection;
    if (empty) return tr;

    const markTypesToRemove = new Set(
        FORMAT_MARK_NAMES.map(n => schema.marks[n]).filter(Boolean)
    );

    if (!markTypesToRemove.size) return tr;

    const tasks = [];
    doc.nodesBetween(from, to, (node, pos) => {
        if (node.marks && node.marks.length) {
            node.marks.some(mark => {
                if (markTypesToRemove.has(mark.type)) {
                    tasks.push({ node, pos, mark });
                }
            });
            return true;
        }
        return true;
    });

    tasks.forEach(job => {
        const { node, mark, pos } = job;
        tr = tr.removeMark(pos, pos + node.nodeSize, mark.type);
    });

    tr = setTextAlign(tr, null);
    tr = setTextLineHeight(tr, null);

    return tr;
}

export { clearMarks };

I hope the above can be of some help to you.

And these three references are from the original author's source code files:
import { setTextAlign } from './text_align';
import { setTextLineHeight } from './line_height';
import CustomStrike from "./paste.js";

**Reply by rudiboot** [reply]
Thanks for that 000yuying I'm going to give it a try

**Comment by drifteaur** [comment]
OK, I tried the above, the minimal version I needed to make it work:

under resources/views/tiptap/tiptap-clean-formatting.blade.php

```
<x-filament-tiptap-editor::button
    label="Clear Formatting"
    action="editor().chain().focus().clearNodes().unsetAllMarks().run()">

    <svg xmlns="http://www.w3.org/2000/svg"
         width="20" height="20" viewBox="0 0 24 24">
        <path fill="currentColor"
              d="M12.651 14.065L11.605 20H9.574l1.35-7.661-7.41-7.41L4.93 3.515 20.485 19.07l-1.414 1.414-6.42-6.42zm-.878-6.535l.27-1.53h-1.8l-2-2H20v2h-5.927L13.5 9.257 11.773 7.53z"/>
    </svg>
    <span class="sr-only">Clear Formatting</span>
</x-filament-tiptap-editor::button>
```

under app/helpers/TipTapClean.php:

```
<?php namespace App\Helpers;

use Tiptap\Core\Node;

class TipTapClean extends Node
{
    public static $name = 'clean';
}
```

In config/filament-tiptap-editor.php:

```
...
    'extensions'                 => [
        [
            'id'     => 'clean',
            'name'   => 'Clean',
            'button' => 'tiptap.tiptap-clean-formatting',
            'parser' => TipTapClean::class,
        ],
    ],
```

Hope it helps
